File: api/testing/assertions.go
package testing

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

// runAssertion runs a single assertion and returns the result.
func (e *TestExecutor) runAssertion(assertion TestAssertion) AssertionResult {
	result := AssertionResult{
		Type: assertion.Type,
	}

	switch assertion.Type {
	case AssertionFileExists:
		result = e.assertFileExists(assertion.Path)
	case AssertionFileNotExists:
		result = e.assertFileNotExists(assertion.Path)
	case AssertionDirExists:
		result = e.assertDirExists(assertion.Path)
	case AssertionDirNotExists:
		result = e.assertDirNotExists(assertion.Path)
	case AssertionFileContains:
		result = e.assertFileContains(assertion.Path, assertion.Contains)
	case AssertionFileNotContains:
		result = e.assertFileNotContains(assertion.Path, assertion.Contains)
	case AssertionFileMatches:
		result = e.assertFileMatches(assertion.Path, assertion.Pattern)
	case AssertionFileEquals:
		result = e.assertFileEquals(assertion.Path, assertion.Value)
	case AssertionOutputEquals:
		result = e.assertOutputEquals(assertion.Block, assertion.Output, assertion.Value)
	case AssertionOutputMatches:
		result = e.assertOutputMatches(assertion.Block, assertion.Output, assertion.Pattern)
	case AssertionOutputExists:
		result = e.assertOutputExists(assertion.Block, assertion.Output)
	case AssertionFilesGenerated:
		result = e.assertFilesGenerated(assertion.MinCount)
	case AssertionScript:
		result = e.assertScript(assertion.Command)
	default:
		result.Passed = false
		result.Message = fmt.Sprintf("unknown assertion type: %s", assertion.Type)
	}

	result.Type = assertion.Type
	return result
}

// resolvePath resolves a path relative to the output directory (workingDir + outputPath).
func (e *TestExecutor) resolvePath(path string) string {
	if filepath.IsAbs(path) {
		return path
	}
	resolvedOutputPath := e.resolveOutputPath()
	return filepath.Join(resolvedOutputPath, path)
}

// assertFileExists checks that a file exists.
func (e *TestExecutor) assertFileExists(path string) AssertionResult {
	fullPath := e.resolvePath(path)
	info, err := os.Stat(fullPath)
	if err != nil {
		if os.IsNotExist(err) {
			return AssertionResult{
				Type:    AssertionFileExists,
				Passed:  false,
				Message: fmt.Sprintf("file does not exist: %s", path),
			}
		}
		return AssertionResult{
			Type:    AssertionFileExists,
			Passed:  false,
			Message: fmt.Sprintf("error checking file: %v", err),
		}
	}
	if info.IsDir() {
		return AssertionResult{
			Type:    AssertionFileExists,
			Passed:  false,
			Message: fmt.Sprintf("path exists but is a directory: %s", path),
		}
	}
	return AssertionResult{
		Type:   AssertionFileExists,
		Passed: true,
	}
}

// assertFileNotExists checks that a file does not exist.
func (e *TestExecutor) assertFileNotExists(path string) AssertionResult {
	fullPath := e.resolvePath(path)
	_, err := os.Stat(fullPath)
	if err != nil {
		if os.IsNotExist(err) {
			return AssertionResult{
				Type:   AssertionFileNotExists,
				Passed: true,
			}
		}
		return AssertionResult{
			Type:    AssertionFileNotExists,
			Passed:  false,
			Message: fmt.Sprintf("error checking file: %v", err),
		}
	}
	return AssertionResult{
		Type:    AssertionFileNotExists,
		Passed:  false,
		Message: fmt.Sprintf("file exists but should not: %s", path),
	}
}

// assertDirExists checks that a directory exists.
func (e *TestExecutor) assertDirExists(path string) AssertionResult {
	fullPath := e.resolvePath(path)
	info, err := os.Stat(fullPath)
	if err != nil {
		if os.IsNotExist(err) {
			return AssertionResult{
				Type:    AssertionDirExists,
				Passed:  false,
				Message: fmt.Sprintf("directory does not exist: %s", path),
			}
		}
		return AssertionResult{
			Type:    AssertionDirExists,
			Passed:  false,
			Message: fmt.Sprintf("error checking directory: %v", err),
		}
	}
	if !info.IsDir() {
		return AssertionResult{
			Type:    AssertionDirExists,
			Passed:  false,
			Message: fmt.Sprintf("path exists but is not a directory: %s", path),
		}
	}
	return AssertionResult{
		Type:   AssertionDirExists,
		Passed: true,
	}
}

// assertDirNotExists checks that a directory does not exist.
func (e *TestExecutor) assertDirNotExists(path string) AssertionResult {
	fullPath := e.resolvePath(path)
	_, err := os.Stat(fullPath)
	if err != nil {
		if os.IsNotExist(err) {
			return AssertionResult{
				Type:   AssertionDirNotExists,
				Passed: true,
			}
		}
		return AssertionResult{
			Type:    AssertionDirNotExists,
			Passed:  false,
			Message: fmt.Sprintf("error checking directory: %v", err),
		}
	}
	return AssertionResult{
		Type:    AssertionDirNotExists,
		Passed:  false,
		Message: fmt.Sprintf("directory exists but should not: %s", path),
	}
}

// assertFileContains checks that a file contains a substring.
func (e *TestExecutor) assertFileContains(path, contains string) AssertionResult {
	fullPath := e.resolvePath(path)
	content, err := os.ReadFile(fullPath)
	if err != nil {
		return AssertionResult{
			Type:    AssertionFileContains,
			Passed:  false,
			Message: fmt.Sprintf("failed to read file: %v", err),
		}
	}

	if strings.Contains(string(content), contains) {
		return AssertionResult{
			Type:   AssertionFileContains,
			Passed: true,
		}
	}

	return AssertionResult{
		Type:    AssertionFileContains,
		Passed:  false,
		Message: fmt.Sprintf("file %s does not contain %q", path, contains),
	}
}

// assertFileNotContains checks that a file does not contain a substring.
func (e *TestExecutor) assertFileNotContains(path, contains string) AssertionResult {
	fullPath := e.resolvePath(path)
	content, err := os.ReadFile(fullPath)
	if err != nil {
		return AssertionResult{
			Type:    AssertionFileNotContains,
			Passed:  false,
			Message: fmt.Sprintf("failed to read file: %v", err),
		}
	}

	if !strings.Contains(string(content), contains) {
		return AssertionResult{
			Type:   AssertionFileNotContains,
			Passed: true,
		}
	}

	return AssertionResult{
		Type:    AssertionFileNotContains,
		Passed:  false,
		Message: fmt.Sprintf("file %s contains %q but should not", path, contains),
	}
}

// assertFileMatches checks that a file matches a regex pattern.
func (e *TestExecutor) assertFileMatches(path, pattern string) AssertionResult {
	fullPath := e.resolvePath(path)
	content, err := os.ReadFile(fullPath)
	if err != nil {
		return AssertionResult{
			Type:    AssertionFileMatches,
			Passed:  false,
			Message: fmt.Sprintf("failed to read file: %v", err),
		}
	}

	re, err := regexp.Compile(pattern)
	if err != nil {
		return AssertionResult{
			Type:    AssertionFileMatches,
			Passed:  false,
			Message: fmt.Sprintf("invalid regex pattern: %v", err),
		}
	}

	if re.Match(content) {
		return AssertionResult{
			Type:   AssertionFileMatches,
			Passed: true,
		}
	}

	return AssertionResult{
		Type:    AssertionFileMatches,
		Passed:  false,
		Message: fmt.Sprintf("file %s does not match pattern %q", path, pattern),
	}
}

// assertFileEquals checks that a file content equals the expected value.
func (e *TestExecutor) assertFileEquals(path, expected string) AssertionResult {
	fullPath := e.resolvePath(path)
	content, err := os.ReadFile(fullPath)
	if err != nil {
		return AssertionResult{
			Type:    AssertionFileEquals,
			Passed:  false,
			Message: fmt.Sprintf("failed to read file: %v", err),
		}
	}

	actual := string(content)
	if actual == expected {
		return AssertionResult{
			Type:   AssertionFileEquals,
			Passed: true,
		}
	}

	return AssertionResult{
		Type:    AssertionFileEquals,
		Passed:  false,
		Message: fmt.Sprintf("file %s content does not equal expected value", path),
	}
}

// assertOutputEquals checks that a block output equals the expected value.
func (e *TestExecutor) assertOutputEquals(blockID, outputName, expected string) AssertionResult {
	outputs, ok := e.blockOutputs[blockID]
	if !ok {
		return AssertionResult{
			Type:    AssertionOutputEquals,
			Passed:  false,
			Message: fmt.Sprintf("block %q has no outputs", blockID),
		}
	}

	actual, ok := outputs[outputName]
	if !ok {
		return AssertionResult{
			Type:    AssertionOutputEquals,
			Passed:  false,
			Message: fmt.Sprintf("block %q has no output %q", blockID, outputName),
		}
	}

	if actual == expected {
		return AssertionResult{
			Type:   AssertionOutputEquals,
			Passed: true,
		}
	}

	return AssertionResult{
		Type:    AssertionOutputEquals,
		Passed:  false,
		Message: fmt.Sprintf("output %s.%s = %q, expected %q", blockID, outputName, actual, expected),
	}
}

// assertOutputMatches checks that a block output matches a regex pattern.
func (e *TestExecutor) assertOutputMatches(blockID, outputName, pattern string) AssertionResult {
	outputs, ok := e.blockOutputs[blockID]
	if !ok {
		return AssertionResult{
			Type:    AssertionOutputMatches,
			Passed:  false,
			Message: fmt.Sprintf("block %q has no outputs", blockID),
		}
	}

	actual, ok := outputs[outputName]
	if !ok {
		return AssertionResult{
			Type:    AssertionOutputMatches,
			Passed:  false,
			Message: fmt.Sprintf("block %q has no output %q", blockID, outputName),
		}
	}

	re, err := regexp.Compile(pattern)
	if err != nil {
		return AssertionResult{
			Type:    AssertionOutputMatches,
			Passed:  false,
			Message: fmt.Sprintf("invalid regex pattern: %v", err),
		}
	}

	if re.MatchString(actual) {
		return AssertionResult{
			Type:   AssertionOutputMatches,
			Passed: true,
		}
	}

	return AssertionResult{
		Type:    AssertionOutputMatches,
		Passed:  false,
		Message: fmt.Sprintf("output %s.%s = %q does not match pattern %q", blockID, outputName, actual, pattern),
	}
}

// assertOutputExists checks that a block output exists.
func (e *TestExecutor) assertOutputExists(blockID, outputName string) AssertionResult {
	outputs, ok := e.blockOutputs[blockID]
	if !ok {
		return AssertionResult{
			Type:    AssertionOutputExists,
			Passed:  false,
			Message: fmt.Sprintf("block %q has no outputs", blockID),
		}
	}

	if _, ok := outputs[outputName]; ok {
		return AssertionResult{
			Type:   AssertionOutputExists,
			Passed: true,
		}
	}

	return AssertionResult{
		Type:    AssertionOutputExists,
		Passed:  false,
		Message: fmt.Sprintf("block %q has no output %q", blockID, outputName),
	}
}

// assertFilesGenerated checks that at least minCount files were generated in the output directory.
func (e *TestExecutor) assertFilesGenerated(minCount int) AssertionResult {
	resolvedOutputPath := e.resolveOutputPath()
	count := 0
	err := filepath.Walk(resolvedOutputPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if !info.IsDir() {
			count++
		}
		return nil
	})

	if err != nil {
		return AssertionResult{
			Type:    AssertionFilesGenerated,
			Passed:  false,
			Message: fmt.Sprintf("failed to walk output directory %q: %s", resolvedOutputPath, err.Error()),
		}
	}

	if count >= minCount {
		return AssertionResult{
			Type:   AssertionFilesGenerated,
			Passed: true,
		}
	}

	return AssertionResult{
		Type:    AssertionFilesGenerated,
		Passed:  false,
		Message: fmt.Sprintf("expected at least %d files generated, got %d", minCount, count),
	}
}

// assertScript runs a custom script command and checks for success.
func (e *TestExecutor) assertScript(command string) AssertionResult {
	duration := e.timeout
	if duration <= 0 {
		duration = 30 * time.Second
	}
	ctx, cancel := context.WithTimeout(context.Background(), duration)
	defer cancel()

	cmd := exec.CommandContext(ctx, "bash", "-c", command)
	cmd.Dir = e.resolveOutputPath()

	// Set environment from session
	execCtx, _ := e.session.ValidateToken(e.getSessionToken())
	if execCtx != nil {
		cmd.Env = execCtx.Env
	}

	err := cmd.Run()
	if err != nil {
		return AssertionResult{
			Type:    AssertionScript,
			Passed:  false,
			Message: fmt.Sprintf("script assertion failed: %v", err),
		}
	}

	return AssertionResult{
		Type:   AssertionScript,
		Passed: true,
	}
}

File: cli/test/assertions.ts
/**
 * Test assertion runners.
 */
import * as fs from "node:fs"
import * as path from "node:path"
import { execFileSync } from "node:child_process"
import type { TestAssertion, AssertionResult } from "./config.ts"

// ---------------------------------------------------------------------------
// Assertion executor
// ---------------------------------------------------------------------------

export interface AssertionContext {
  /** Absolute path to the output directory (workingDir + outputPath). */
  outputDir: string
  /** Block outputs collected during test execution. */
  blockOutputs: Map<string, Map<string, string>>
  /** Session env vars (as KEY=VALUE strings). */
  sessionEnv: string[]
  /** Timeout in ms for script assertions. */
  timeout: number
}

/** Run a single assertion and return the result. */
export function runAssertion(
  assertion: TestAssertion,
  ctx: AssertionContext,
): AssertionResult {
  switch (assertion.type) {
    case "file_exists": return assertFileExists(assertion.path!, ctx)
    case "file_not_exists": return assertNotExists("file", assertion.path!, ctx)
    case "dir_exists": return assertDirExists(assertion.path!, ctx)
    case "dir_not_exists": return assertNotExists("dir", assertion.path!, ctx)
    case "file_contains": return assertFileContains(assertion.path!, assertion.contains!, ctx)
    case "file_not_contains": return assertFileNotContains(assertion.path!, assertion.contains!, ctx)
    case "file_matches": return assertFileMatches(assertion.path!, assertion.pattern!, ctx)
    case "file_equals": return assertFileEquals(assertion.path!, assertion.value!, ctx)
    case "output_equals": return assertOutputEquals(assertion.block!, assertion.output!, assertion.value ?? "", ctx)
    case "output_matches": return assertOutputMatches(assertion.block!, assertion.output!, assertion.pattern!, ctx)
    case "output_exists": return assertOutputExists(assertion.block!, assertion.output!, ctx)
    case "files_generated": return assertFilesGenerated(assertion.min_count ?? 0, ctx)
    case "script": return assertScript(assertion.command!, ctx)
    default:
      return { type: assertion.type, passed: false, message: `Unknown assertion type: ${assertion.type}` }
  }
}

// ---------------------------------------------------------------------------
// Path resolution
// ---------------------------------------------------------------------------

function resolvePath(relPath: string, ctx: AssertionContext): string {
  if (path.isAbsolute(relPath)) return relPath
  return path.join(ctx.outputDir, relPath)
}

// ---------------------------------------------------------------------------
// File assertions
// ---------------------------------------------------------------------------

function assertFileExists(filePath: string, ctx: AssertionContext): AssertionResult {
  const fullPath = resolvePath(filePath, ctx)
  try {
    const stat = fs.statSync(fullPath)
    if (stat.isDirectory()) {
      return { type: "file_exists", passed: false, message: `Path exists but is a directory: ${filePath}` }
    }
    return { type: "file_exists", passed: true }
  } catch (e: unknown) {
    if ((e as NodeJS.ErrnoException).code === "ENOENT") {
      return { type: "file_exists", passed: false, message: `File does not exist: ${filePath}` }
    }
    return { type: "file_exists", passed: false, message: `Error checking file: ${e}` }
  }
}

function assertNotExists(
  kind: "file" | "dir",
  targetPath: string,
  ctx: AssertionContext,
): AssertionResult {
  const type = kind === "file" ? "file_not_exists" : "dir_not_exists"
  const noun = kind === "file" ? "File" : "Directory"
  const lowerNoun = kind === "file" ? "file" : "directory"
  const fullPath = resolvePath(targetPath, ctx)
  try {
    fs.statSync(fullPath)
    return { type, passed: false, message: `${noun} exists but should not: ${targetPath}` }
  } catch (e: unknown) {
    if ((e as NodeJS.ErrnoException).code === "ENOENT") {
      return { type, passed: true }
    }
    return { type, passed: false, message: `Error checking ${lowerNoun}: ${e}` }
  }
}

function assertDirExists(dirPath: string, ctx: AssertionContext): AssertionResult {
  const fullPath = resolvePath(dirPath, ctx)
  try {
    const stat = fs.statSync(fullPath)
    if (!stat.isDirectory()) {
      return { type: "dir_exists", passed: false, message: `Path exists but is not a directory: ${dirPath}` }
    }
    return { type: "dir_exists", passed: true }
  } catch (e: unknown) {
    if ((e as NodeJS.ErrnoException).code === "ENOENT") {
      return { type: "dir_exists", passed: false, message: `Directory does not exist: ${dirPath}` }
    }
    return { type: "dir_exists", passed: false, message: `Error checking directory: ${e}` }
  }
}

function assertFileContains(filePath: string, substring: string, ctx: AssertionContext): AssertionResult {
  const fullPath = resolvePath(filePath, ctx)
  try {
    const content = fs.readFileSync(fullPath, "utf-8")
    if (content.includes(substring)) {
      return { type: "file_contains", passed: true }
    }
    return { type: "file_contains", passed: false, message: `File ${filePath} does not contain "${substring}"` }
  } catch (e: unknown) {
    return { type: "file_contains", passed: false, message: `Failed to read file: ${e}` }
  }
}

function assertFileNotContains(filePath: string, substring: string, ctx: AssertionContext): AssertionResult {
  const fullPath = resolvePath(filePath, ctx)
  try {
    const content = fs.readFileSync(fullPath, "utf-8")
    if (!content.includes(substring)) {
      return { type: "file_not_contains", passed: true }
    }
    return { type: "file_not_contains", passed: false, message: `File ${filePath} contains "${substring}" but should not` }
  } catch (e: unknown) {
    return { type: "file_not_contains", passed: false, message: `Failed to read file: ${e}` }
  }
}

function assertFileMatches(filePath: string, pattern: string, ctx: AssertionContext): AssertionResult {
  const fullPath = resolvePath(filePath, ctx)
  try {
    const content = fs.readFileSync(fullPath, "utf-8")
    const re = new RegExp(pattern)
    if (re.test(content)) {
      return { type: "file_matches", passed: true }
    }
    return { type: "file_matches", passed: false, message: `File ${filePath} does not match pattern "${pattern}"` }
  } catch (e: unknown) {
    if (e instanceof SyntaxError) {
      return { type: "file_matches", passed: false, message: `Invalid regex pattern: ${e.message}` }
    }
    return { type: "file_matches", passed: false, message: `Failed to read file: ${e}` }
  }
}

function assertFileEquals(filePath: string, expected: string, ctx: AssertionContext): AssertionResult {
  const fullPath = resolvePath(filePath, ctx)
  try {
    const content = fs.readFileSync(fullPath, "utf-8")
    if (content === expected) {
      return { type: "file_equals", passed: true }
    }
    return { type: "file_equals", passed: false, message: `File ${filePath} content does not equal expected value` }
  } catch (e: unknown) {
    return { type: "file_equals", passed: false, message: `Failed to read file: ${e}` }
  }
}

// ---------------------------------------------------------------------------
// Output assertions
// ---------------------------------------------------------------------------

function assertOutputEquals(
  blockId: string,
  outputName: string,
  expected: string,
  ctx: AssertionContext,
): AssertionResult {
  const outputs = ctx.blockOutputs.get(blockId)
  if (!outputs) {
    return { type: "output_equals", passed: false, message: `Block "${blockId}" has no outputs` }
  }
  const actual = outputs.get(outputName)
  if (actual === undefined) {
    return { type: "output_equals", passed: false, message: `Block "${blockId}" has no output "${outputName}"` }
  }
  if (actual === expected) {
    return { type: "output_equals", passed: true }
  }
  return { type: "output_equals", passed: false, message: `output ${blockId}.${outputName} = "${actual}", expected "${expected}"` }
}

function assertOutputMatches(
  blockId: string,
  outputName: string,
  pattern: string,
  ctx: AssertionContext,
): AssertionResult {
  const outputs = ctx.blockOutputs.get(blockId)
  if (!outputs) {
    return { type: "output_matches", passed: false, message: `Block "${blockId}" has no outputs` }
  }
  const actual = outputs.get(outputName)
  if (actual === undefined) {
    return { type: "output_matches", passed: false, message: `Block "${blockId}" has no output "${outputName}"` }
  }
  try {
    const re = new RegExp(pattern)
    if (re.test(actual)) {
      return { type: "output_matches", passed: true }
    }
    return { type: "output_matches", passed: false, message: `output ${blockId}.${outputName} = "${actual}" does not match pattern "${pattern}"` }
  } catch {
    return { type: "output_matches", passed: false, message: `Invalid regex pattern: ${pattern}` }
  }
}

function assertOutputExists(
  blockId: string,
  outputName: string,
  ctx: AssertionContext,
): AssertionResult {
  const outputs = ctx.blockOutputs.get(blockId)
  if (!outputs) {
    return { type: "output_exists", passed: false, message: `Block "${blockId}" has no outputs` }
  }
  if (outputs.has(outputName)) {
    return { type: "output_exists", passed: true }
  }
  return { type: "output_exists", passed: false, message: `Block "${blockId}" has no output "${outputName}"` }
}

// ---------------------------------------------------------------------------
// Files generated assertion
// ---------------------------------------------------------------------------

function assertFilesGenerated(minCount: number, ctx: AssertionContext): AssertionResult {
  let count = 0
  try {
    count = countFiles(ctx.outputDir)
  } catch (e: unknown) {
    return { type: "files_generated", passed: false, message: `Failed to walk output directory "${ctx.outputDir}": ${e}` }
  }

  if (count >= minCount) {
    return { type: "files_generated", passed: true }
  }
  return { type: "files_generated", passed: false, message: `Expected at least ${minCount} files generated, got ${count}` }
}

export function countFiles(dir: string): number {
  if (!fs.existsSync(dir)) return 0
  let count = 0
  const entries = fs.readdirSync(dir, { withFileTypes: true })
  for (const entry of entries) {
    if (entry.isDirectory()) {
      count += countFiles(path.join(dir, entry.name))
    } else {
      count++
    }
  }
  return count
}

// ---------------------------------------------------------------------------
// Script assertion
// ---------------------------------------------------------------------------

/**
 * Run a script command as a bash assertion.
 * Uses execFileSync with bash to avoid shell injection risks — the command
 * string is passed as a single argument to bash -c, not interpolated into
 * a shell invocation.
 */
function assertScript(command: string, ctx: AssertionContext): AssertionResult {
  try {
    execFileSync("/bin/bash", ["-c", command], {
      cwd: ctx.outputDir,
      env: envListToRecord(ctx.sessionEnv),
      timeout: Math.min(ctx.timeout, 30000),
      stdio: "pipe",
    })
    return { type: "script", passed: true }
  } catch (e: unknown) {
    return { type: "script", passed: false, message: `Script assertion failed: ${e}` }
  }
}

export function envListToRecord(envList: string[]): Record<string, string> {
  const result: Record<string, string> = {}
  for (const entry of envList) {
    const idx = entry.indexOf("=")
    if (idx >= 0) {
      result[entry.slice(0, idx)] = entry.slice(idx + 1)
    }
  }
  return result
}
